src/core/commandRunner/CommandRunner.ts

```diff
@@ -0,0 +1,158 @@
+
+    this.runCommandBackwardCompatible =
+      this.runCommandBackwardCompatible.bind(this);
+
+    this.disposables.push(
+      vscode.commands.registerCommand(
+        "cursorless.command",
+        this.runCommandBackwardCompatible
+      )
+    );
+  }
+
+  async runCommand(commandArgument: CommandArgument) {
+    const {
+      spokenForm,
+      action: inputActionName,
+      targets: inputPartialTargets,
+      extraArgs: inputExtraArgs,
+      usePrePhraseSnapshot = false,
+    } = commandArgument;
+
+    try {
+      const { actionName, partialTargets, extraArgs } =
+        canonicalizeAndValidateCommand(
+          inputActionName,
+          inputPartialTargets,
+          inputExtraArgs ?? []
+        );
+
+      const readableHatMap = await this.graph.hatTokenMap.getReadableMap(
+        usePrePhraseSnapshot
+      );
+
+      console.debug(`commandArgument:`);
+      console.debug(JSON.stringify(commandArgument, null, 3));
+
+      const action = this.graph.actions[actionName];
+
+      if (action == null) {
+        throw new Error(`Unknown action ${actionName}`);
+      }
+
+      const targets = inferFullTargets(
+        partialTargets,
+        action.getTargetPreferences(...extraArgs)
+      );
+
+      const processedTargetsContext: ProcessedTargetsContext = {
+        currentSelections:
+          vscode.window.activeTextEditor?.selections.map((selection) => ({
+            selection,
+            editor: vscode.window.activeTextEditor!,
+          })) ?? [],
+        currentEditor: vscode.window.activeTextEditor,
+        hatTokenMap: readableHatMap,
+        thatMark: this.thatMark.exists() ? this.thatMark.get() : [],
+        sourceMark: this.sourceMark.exists() ? this.sourceMark.get() : [],
+        getNodeAtLocation: this.getNodeAtLocation,
+      };
+
+      const selections = processTargets(processedTargetsContext, targets);
+
+      if (this.testCaseRecorder.active) {
+        const context = {
+          targets,
+          thatMark: this.thatMark,
+          sourceMark: this.sourceMark,
+          hatTokenMap: readableHatMap,
+          spokenForm,
+        };
+        await this.testCaseRecorder.preCommandHook(commandArgument, context);
+      }
+
+      const {
+        returnValue,
+        thatMark: newThatMark,
+        sourceMark: newSourceMark,
+      } = await action.run(selections, ...extraArgs);
+
+      this.thatMark.set(newThatMark);
+      this.sourceMark.set(newSourceMark);
+
+      if (this.testCaseRecorder.active) {
+        await this.testCaseRecorder.postCommandHook(returnValue);
+      }
+
+      return returnValue;
+    } catch (e) {
+      this.testCaseRecorder.commandErrorHook();
+      const err = e as Error;
+      vscode.window.showErrorMessage(err.message);
+      console.debug(err.message);
+      console.debug(err.stack);
+      throw err;
+    }
+  }
+
+  private runCommandBackwardCompatible(
+    spokenFormOrCommandArgument: string | CommandArgument,
+    ...rest: unknown[]
```

Comment:
Here's the ugliness that allows us to handle either old or new command arguments

Reply:
I think that works fine